javalib/src/main/scala/java/util/Base64.scala

```diff
@@ -0,0 +1,486 @@
+        // scalastyle:off return
+        while (src.hasRemaining) {
+          if (srcBuffer.hasRemaining) {
+            val int = src.get & 0xff
+            table(int) match {
+              case -2 => return
+              case -1 if ignoreInvalid =>
+              case -1 => throw new IllegalArgumentException("Illegal base64 character " + Integer.toString(int, 16))
+              case i  => srcBuffer.put(i)
+            }
+          } else inputData()
+        }
+        // scalastyle:on return
+      }
+      iterate()
+      // end or padding
+      srcBuffer.flip()
+      inputData()
+      while(src.hasRemaining) {
+        val int = src.get & 0xff
+        val value = table(int)
+        if (value != -2 && (!ignoreInvalid || value > 0))
+          throw new IllegalArgumentException(s"Input byte array has incorrect ending byte at $int")
+      }
+      dst.position
+    }
+
+    private def dstRequiredLength(src: Array[Byte], offset: Int, length: Int): Int = {
+      // scalastyle:off return
+      if (length == 0) return 0
+      // scalastyle:on return
+
+      var validBytes = 0
+      var padding = 0
+      if (ignoreInvalid) {
+        for (i <- offset until length) {
+          if (table(src(i) & 0xff) >= 0)
+            validBytes += 1
+        }
+      } else {
+        validBytes = length-offset
+        /* We check the end for padding and compute the length from there. This is ok, if the rest contains garbage
+        we'll have written something before throwing but the spec says "If the input byte array is not in valid Base64
+        encoding scheme then some bytes may have been written to the output byte array before IllegalArgumentException
+        is thrown."*/
+        if (src(length-1) == '=') {
+          padding += 1
+          if (src(length-2) == '=')
+          padding += 1
+        }
+      }
+      dstMaxLength(validBytes) - (if (validBytes % 4 == 0) padding else 4 - (validBytes % 4))
+    }
+
+    /** compute the destination length solely based on the source length length without knowing about padding */
+    private def dstMaxLength(srcLength: Int): Int = (srcLength+3) / 4 * 3
+
+  }
+
+  private class DecodingInputStream(in: InputStream, table: Array[Int], ignoreInvalid: Boolean)
+    extends FilterInputStream(in) {
+
+    final val DecodeState18 = 0
+    final val DecodeState12 = 1
+    final val DecodeState14 = 2
+    final val DecodeState16 = 3
+
+    private val oneBuf = new Array[Byte](1)
+
+    private var closed = false
+    private var eof = false
+    private var out = 0
+    private var shift = DecodeState18
+
+    override def read(): Int = if (read(oneBuf, 0, 1) == -1) -1 else oneBuf(0) & 0xff
+
+    override def read(b: Array[Byte], off: Int, len: Int): Int = {
+      // scalastyle:off return
+      if (closed) throw new IOException("Stream is closed")
+      if (off < 0 || len < 0 || len > b.length - off) throw new IndexOutOfBoundsException
+      if (eof) return -1
```

Comment:
Consider moving this after the inner `def`s, as this will allow you to get rid of the `return`:
```scala
if (eof) {
  -1
} else {
  iterate()
  written
}
```
`return`s are *exceedingly* rare in this codebase. We have only 17 `return`s out of 120 kLoC. We only use them when we have no other choice.